[WIP] Generalize AutoYaST verification of cloned profile

jknphy opened the pull request:
Generalize AutoYaST verification of cloned profile.  Added functionality to evaluate expression so we can count number of nodes, but still default is to search by node to detect repetitions. Full list of error is displayed, not stopping in the first one failing.

- Related ticket: https://progress.opensuse.org/issues/60002
- Job Group MR: https://gitlab.suse.de/qsf-y/qa-sle-functional-y/merge_requests/92
- Verification run:
  - [autoyast_disk_as_pv](http://rivera-workstation.suse.cz/tests/680)
  - [autoyast_disk_as_md_member](http://rivera-workstation.suse.cz/tests/681)
  - [autoyast_btrfs](http://rivera-workstation.suse.cz/tests/684)
  - [autoyast_multi_btrfs](http://rivera-workstation.suse.cz/tests/687)

Note 1: For future I guess we can include in yaml possibility to add some base xpath or `$context_node` with some grouping so we can pass it to the [library](https://metacpan.org/pod/XML::LibXML::XPathContext) and shorten current xpaths.
Note 2: [example of list of errors](http://rivera-workstation.suse.cz/tests/661#step/verify_cloned_profile/12)

jknphy commented on schedule/yast/autoyast_disk_as_md_member.yaml:
```diff
@@ -29,6 +29,12 @@ test_data:
-    raid_type: '//ns:partitioning/ns:drive[ns:device="/dev/md/0"]/ns:raid_options/ns:raid_type'
+  cloned_profile:
+    raid_level:
+      xpath: '//ns:partitioning/ns:drive[ns:device="/dev/md/0"]/ns:raid_options/ns:raid_type'
```
I've just realized checking existing code in the repo, that I can read this as a list, so I can avoid to have this keys `raid_level`, etc, which would simplify the yaml a lot.

jknphy replied:
done!

b10n1k commented on lib/xml_utils.pm:
```diff
@@ -51,26 +51,36 @@ Verifies that node by given XPath is unique and has expected value. C<%args> is
+    if ($args{search_by_value}) {
+        my $value = $args{xpc}->findvalue($args{xpath});
+        if ($value ne $args{expected_value}) {
+            return "Unexpected value for xpath $args{xpath}. Expected: '$args{expected_value}', got: '$value'";
```
"Generated autoinst.xml" would be useful here to give some extra information where the error come from.

jknphy replied:
I kept this message and the logic below.

b10n1k replied:
I see. resolve this conversation

b10n1k commented on lib/xml_utils.pm:
```diff
@@ -51,26 +51,36 @@ Verifies that node by given XPath is unique and has expected value. C<%args> is
+        my @nodes = find_nodes(%args);
+        if (scalar @nodes == 0) {
+            return "Expected xpath $args{xpath} didn't return any node";
+        }
```
i dont know if we follow some convention here but if not (still can applied though with another way!) the perl way is more readable 
return "Expected xpath $args{xpath} didn't return any node" if (scalar @nodes == 0);

jknphy replied:
No, in this case is not more readable because the following checks are not in postfix way and the message are pretty long, I would need to put all the message in variables which is even worst... Besides that I'm keeping same structure of previous code, not doing this part from scratch.

jknphy replied:
but I agree with you that those `if/else` looks like could be done in a different way, perhaps having some hash for the message, not sure, ...

b10n1k replied:
no need to make it complicated. was just a suggestion. resolved

b10n1k commented on lib/xml_utils.pm:
```diff
@@ -51,26 +51,36 @@ Verifies that node by given XPath is unique and has expected value. C<%args> is
+        # Verify that there is node found by xpath and it's single one
+        if (scalar @nodes != 1) {
+            return "Generated autoinst.xml contains unexpected number of nodes for xpath: $args{xpath}. Found: " . scalar @nodes . ", expected: 1.";
+        }
```
same here? https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/9058/files#r353295084

jknphy replied:
same as above.

b10n1k commented on lib/xml_utils.pm:
```diff
@@ -51,26 +51,36 @@ Verifies that node by given XPath is unique and has expected value. C<%args> is
+        }
+        if ($nodes[0]->to_literal ne $args{expected_value}) {
+            return "Unexpected value for xpath $args{xpath}. Expected: '$args{expected_value}', got: '$nodes[0]'";
+        }
```
And here same here? https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/9058/files#r353295084

jknphy replied:
In this case is not even recommended, because we have more than one condition...and the postfix notation is even less clear.

b10n1k commented on lib/xml_utils.pm:
```diff
@@ -87,7 +97,7 @@ sub find_nodes {
     my $nodeset = $args{xpc}->findnodes($args{xpath});
     for my $node ($nodeset->get_nodelist) {
-        print $node->to_literal;
+        diag $node->to_literal;
```
why diag here?

jknphy replied:
I'm not sure if helped me to debug something in some point, I cannot recall, but in general we should use `diag` instead of `print` because it is more reliable, it is part of os-autoinst.

b10n1k replied:
i didnt know that there is diag implementation in os-autoinst. i thought it was in the test module. from the documentation "Prints a diagnostic message which is guaranteed not to interfere with test output. Like print @diagnostic_message is simply concatenated together." As i am not expert in perl i am leaving it for others to decide(or you). out of curiosity, pm a link of its output if you want

jknphy replied:
Now I'm thinking, I have added another `diag` when searching by value, so now we can see printed in the logs all values are evaluated, not only when searching by node like in [here](http://rivera-workstation.suse.cz/tests/687/file/autoinst-log.txt) when you search `[2019-12-03T15:25:16.141 CET] [debug] gpt` you can see them. Thanks for taking a look, I would have not noticed that. I think before was not printed for some reason, also this thing is running in the worker. Let's leave it like that, seems useful.

b10n1k commented on tests/autoyast/verify_cloned_profile.pm:
```diff
@@ -0,0 +1,38 @@
+    my $errors;
+    my $test_data      = get_test_suite_data();
+    my $cloned_profile = $test_data->{cloned_profile};        # get test data related with ay profile
+    my $xpc            = get_xpc(init_autoyast_profile());    # get XPathContext
```
$xpc is not descriptive. i guess neither get_xpc but as you dont introduce this, it can be kept as it is. but the variable would improve a lot the readability

jknphy replied:
It is enough descriptive because it looks very much like some copy/paste name convention from examples of use of a library, so it is easy too google it: https://metacpan.org/pod/XML::LibXML::XPathContext. Besides there is a comment already there just in case.

b10n1k replied:
i dont get that. yes the comment is here and helps but the variable is used elsewhere. and using a slightly more descriptive name would take away the need of comments. obviously not a problem if you dont want to change it as we are not restrict in those kind of things. thanks for the respond

jknphy replied:
my point is that when I see in the code something that is not meaningful like xpc, that is kind of the acronysm, I always think that is the standard way to name the variable for the use of a library (I have seen in other languages same thing), so in fact it helps when you google for more info to use it. I would not even need the comments, but you're right, this is an opinion.

jknphy commented:
> I will just leave comment about the approach. Regardless of the code changes, I would expect that from test data we can understand what is tested and easily come up with test.

It is the final goal, sure, simple code in the test module driven by the data, I totally agree.

> Your approach allows to test any given xml, but it will be hell to investigate failures. On top of it, it's not clear which xpath tests what (except generic statement like we expect 6 nodes).

It is quite trivial investigate failures, but it is too generic as you correctly pointed. This expression (like count) are introduced to not cut test coverage, they were there but done in Perl and I think I added a few more. It would be very easy to add a "message" for each check so the error could be more like: "4 hd were expected plus two multidevice btrfs" or something like that. Besides that using MAKESTNAPSHOTS, which seems to work for me all the time for ay tests it is straightforward to debug the test (but I'm thinking as a developer here more than as a tester which is not so good).

> What I would expect here is that we have some structure to tests AY profile and not generic xml.
> Like we provide disk, and specify expectations for it, like mount points, partitions, subvolumes for given partition, etc.
> 
> What scares me most is complexity of xpath expressions which we repeat many times and amount of checks which is hard to understand. Feels like raid test we had tested with regular expressions.

Repetitions is due to it is not implemented some other field to group areas of the xml and start from it, so it would be shortened a lot. Regular expression are more tricky to understand, in general this is just to follow a path on the xml (but you're right that someone could introduce a wicked expression) but I got your point here as I spent 3 hours not knowing why this was not working due to I was looking the the wrong profile...

Anyway I just wanted to give you more detail about why I implemented this solution, but **I will try a second approach and we can compare**, so I will have in mind some design where we have simple data , some library/layer in the middle to do the heavy work and be able to write simple test. 

I guess we can discard to compare .xml (input profile vs cloned profile) due to this would not give us to much info except debuging xmls, so I will try to mimic structure of xml in yaml and see how it goes.

In case of a second approach would not be enough satisfying , let's keep in mind that this improvement avoid to have multiple test modules to test ay cloned profiles, so basically unify what we already had.

Thanks for the feedback.

jknphy commented:
Let's close this approach as #9168 was merged.
